### ISSUE TITLE
Suggestion - UI Revamp (based on Paint.NET)

### ISSUE THREAD

**Issue by david-szabo97** [piskelapp/piskel#540]
I was thinking about a rework of the UI. There are many paint/photo editor software and for example I really like the simpleness of the Paint.NET (http://www.getpaint.net/index.html)

The UI is pretty simple and it's really user-friendly (atleast for me). Every window can be moved and the extra features can be accessed through the menu bar. Also, you can edit multiple files within one instance.

![](https://puu.sh/qW3bk/dccc8cff09.png)

I think the same concept / layout could be used at Piskel too and it would be easier to make it compatible with lower resolutions. (mobile)

Though I really like the current layout, it's not so extensible, really dependent on the resolution and static / not customizable.

Frames, Layers, Palettes, Tools, Primary / Secondary color, Preview -> separate window (movable)
Settings -> menu bar (opens a dialog box rather than a drawer)

All of this in the dark theme with later additional support for a light theme.

I could tackle the revamp after creating some prototypes and posting it here for feedback.

**Comment by caleybrock** [comment]
Adding myself to this so I can follow along.

We've added a css file to [our fork of Piskel](https://github.com/code-dot-org/piskel/pull/11) to give it a different style. We haven't changed any dom elements to do this, and our hope was that we'd still be able to pull updates from here without too many conflicts. Depending on how you make these changes, I'll probably still try to keep the code-dot-org fork in sync.

**Comment by david-szabo97** [comment]
I achieved the dynamic UI with very minimal changes to the DOM and CSS.

I think the framelist should stay on the left with a button to hide/show and/or dock/undock.

I was just playing around. The header will be replaced with a menu bar for accessing some other features. The right side settings drawers will be replaced with dialogs and the actions will be moved to the menu bar.

What do you guys think? Should I continue working on it? @juliandescottes 

![](https://puu.sh/qYZiX/bea05778b4.png)

**Comment by epicabsol** [comment]
Wow! I like the improvements, although I would like to be able to dock all the dialogs to the edges of the workspace because I don't like floaty windows getting in the way of the my drawing. Without docking, their placement just feels too arbitrary and cluttered. However, being able to arrange them however I like would be useful, especially for varying screen sizes. Also, I like the idea of being able to hide / show windows. I would prefer collapsing dialogs to just the dialog title instead of hiding the entire dialog.

Should window layout be stored in the browser, in the server account, or just destroyed with the session?

Also a minor nitpick: Could you make the dialog shadows be transparent black? Grey shadows on top of a darker background always stick out to me.
Wait - those aren't shadows, those are oddly thick dialog borders. My bad. The border color isn't very distinct from the background of the drawing area, and should shadows be added?

Keep up the great work!

**Comment by juliandescottes** [comment]
👍 

Thanks for the proposition. I think overall having more customization for the UI is the way to go.

tldr: this is going in the right direction, but there's a lot to discuss ! :)

One concern with this approach is the default behavior. We need to make sure that by default, the UI looks nice, works well when resizing the window, when adding layers & colors. How will this look like for a first time user?

With floating panels, I think some panels should also be reworked:
- move the color pickers with the color palettes (to actually make a panel similar to the colors panel of paint . net)
- palette and layer panels should be resizable vertically (and no longer resize automatically depending on their content)

Floating panels would also enable some other nice features in the future
- preview panel could be resizable (and we could remove the current preview popup, which is often inconvenient)
- tools (& transform tools) panels could also be resized to have more horizontal icon layouts

I also agree with @epicabsol , it would be nice to pair this with a good docking mechanism. This means docking on the edges of the viewport, as well as between panels. 

> Should window layout be stored in the browser, in the server account, or just destroyed with the session?

For now we can only store it in the browser, later on we could think about saving user preferences on the piskel account.

Some minor feedback on the current approach:
- keyboard shortcut icon (bottom left) will have to move(to a menu maybe, otherwise it will overlap with the framelist)
- cursor coordinates (bottom right) are no longer visible. Needs add a background to make sure they're always readable?
- zoomed out margin color (lightgrey) should maybe be the same color as the background color now ?

**Comment by caleybrock** [comment]
While I like being able to move around the tools, it'd be great to have a flag or option to be able to lock tools in place (from the code). Code.org is using piskel in Gamelab, which a tool for middle/high school students to program their own games. I'd be worried that if we used this UI for our use case, students could get in a bad state and have trouble getting back. Also schools tend to have much smaller screens with lower resolution and we wouldn't want anything to be more complicated for them.

**Comment by juliandescottes** [comment]
@david-szabo97 Hi! Are you still interested in working on this? 

I think this would really be a great feature, provided we can make it optional and ensure the original layout still works the same as today.

**Comment by david-szabo97** [comment]
@juliandescottes Hi! It would be great to work on this but I'm kinda busy.

If I will have time for this and this is still undone then I can take it.
